### DISCUSSION TITLE
Using `cs` for one component silently doubles the memory of everything

### DISCUSSION THREAD

**Discussion by joanibal** [OpenMDAO/OpenMDAO#3207] [General]
Hello,

I recently stumbled upon an aspect of OpenMDAO that I found surprising/confusing.  I was trying to run a larger case in MPhys but kept running out of memory. Confusingly it was trying to allocate an extra complex part for all my large CFD variables. That is when I found this section of the code.
https://github.com/OpenMDAO/OpenMDAO/blob/0fec22f266b7d839907b294d4cd0938cc39889e3/openmdao/core/group.py#L1125-L1132
It turns out an extra little component that I had set to use 'cs' was doubled the memory footprint of everything. 

I see that this is mentioned in the [docs](https://openmdao.org/newdocs/versions/latest/advanced_user_guide/complex_step.html). 
>"When complex step is used somewhere in an OpenMDAO model, OpenMDAO must allocate sufficient memory to contain a complex version of the nonlinear vector" 

However, I think more clarification in the docs on the meaning of "the nonlinear vector" and the implication (doubling the memory of everything) would be helpful. On that note, a warning could also help (if a subsystem needs a complex part and `force_alloc_complex=False`). I could see there being a few tricky situations where ones connects MPhys with another OpenMDAO library that uses `cs` without knowing.

**Comment by robfalck** [comment]
We can look at making it possible for other components to do the "local CS" that ExecComp does.
Another way to do this might be to use a SubmodelComp, where the machinery of the MAUD architecture (inputs, outputs, residuals, etc) is firewalled at the sub-problem level.

**Comment by joanibal** [comment]
Thanks Rob. A "local CS" would be very useful for me. In the meantime I'll avoid CS in favor of hand AD for everything.
